zrml/swaps/src/lib.rs

```diff
@@ -1208,6 +1238,113 @@ mod pallet {
+                extra_weight_per_pool,
+            )
+        }
+
+        // Apply a `mutation` to all pools cached for arbitrage (but at most `pool_count` many) and
+        // return the actual weight consumed.
+        pub(crate) fn apply_to_cached_pools<F>(
+            pool_count: u32,
+            mutation: F,
+            max_weight_per_pool: Weight,
+        ) -> Weight
+        where
+            F: Fn(PoolId) -> Result<Weight, DispatchError>,
+        {
+            let mut total_weight = T::WeightInfo::apply_to_cached_pools_noop(pool_count);
+            for (index, (pool_id, _)) in PoolsCachedForArbitrage::<T>::drain().enumerate() {
+                // The mutation should never fail, but if it does, we just assume we
+                // consumed all the weight and rollback the pool.
+                let _ = with_transaction(|| match mutation(pool_id) {
+                    Err(err) => {
+                        log::warn!(
+                            "Arbitrage unexpectedly failed on pool {:?} with error: {:?}",
+                            pool_id,
+                            err,
+                        );
+                        total_weight = total_weight.saturating_add(max_weight_per_pool);
+                        TransactionOutcome::Rollback(err.into())
+                    }
+                    Ok(weight) => {
+                        total_weight = total_weight.saturating_add(weight);
+                        TransactionOutcome::Commit(Ok(()))
+                    }
+                });
+                if index.saturating_add(1) == (pool_count as usize) {
+                    break;
+                }
+            }
+            total_weight
+        }
+
+        // Execute arbitrage on a single pool.
+        pub(crate) fn execute_arbitrage(
+            pool_id: PoolId,
+            max_iterations: usize,
+        ) -> Result<Weight, DispatchError> {
+            let pool = Self::pool_by_id(pool_id)?;
+            let pool_account = Self::pool_account_id(pool_id);
+            let balances = pool
+                .assets
+                .iter()
+                .map(|a| (*a, T::AssetManager::free_balance(*a, &pool_account)))
+                .collect::<BTreeMap<_, _>>();
+            let asset_count = pool.assets.len() as u32;
+            let outcome_tokens = pool.assets.iter().filter(|a| **a != pool.base_asset);
+            let total_spot_price = pool.calc_total_spot_price(&balances)?;
+
+            if total_spot_price > BASE.saturating_add(ARBITRAGE_THRESHOLD) {
+                let (amount, iteration_count) =
+                    pool.calc_arbitrage_amount_mint_sell(&balances, max_iterations)?;
+                T::AssetManager::withdraw(pool.base_asset, &pool_account, amount)?;
+                for t in outcome_tokens {
+                    T::AssetManager::deposit(*t, &pool_account, amount)?;
+                }
+                Self::deposit_event(Event::ArbitrageMintSell(pool_id, amount));
```

Comment:
Yes. And now that you say this, I notice I've made a huge mistake here: When we create the outcome tokens, we shouldn't _withdraw_ the ZTG, but rather _transfer_ it from the liquidity pool to the prize pool. Otherwise, some of our outcome tokens are not collateralized.

Reply:
423e311131371757d7c7fc303b61a47b639d61d4